### PR TITLE
fix potential race condition

### PR DESCRIPTION
**Pull request by MUzairS15**
**Notes for Reviewers**

This PR fixes a potential race condition which may occur. 

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

File: server/internal/graphql/resolver/configuration.go [hunk 1]
```diff
@@ -22,6 +22,7 @@ func (r *Resolver) subscribeConfiguration(ctx context.Context, provider models.P
 	chp <- struct{}{}
 	cha <- struct{}{}
 	chf <- struct{}{}
+	key := len(r.Config.ConfigurationChannel.ApplicationsChannel)
```
Comment (Revolyssup) [inline server/internal/graphql/resolver/configuration.go:25]: @MUzairS15 Given the fact that multiple instances of this function will be called for each client, here this key will introduce another race condition. 
Imagine this: Two clients connect at approx the same time. In both instances of this function, len will turn out to be the same and hence two different subscription will get the same key.
Use UUID or something like that.

**Review by Revolyssup — CHANGES_REQUESTED** [review]

**Comment by Azanul** [comment]
@MUzairS15 I couldn't see a race. Will you please point it out for me.

**Comment by MUzairS15** [comment]
> @MUzairS15 I couldn't see a race. Will you please point it out for me.

Basically, here in this function https://github.com/meshery/meshery/blob/1949742026f4828ddbd3c5bf90d5008b51550c78/server/helpers/utils/utils.go#L112
we are expecting value and, based on this, deciding whether the channel is closed or not. https://github.com/meshery/meshery/pull/7309/files#diff-c9ba355fc474472a3b057eed347eb447b73f8ae534b2277c78f1767ac04a8ad0L33 and here we are publishing to the same channel.

Secondly, we were closing channel on the receiver's side.

**Comment by Azanul** [comment]
> > @MUzairS15 I couldn't see a race. Will you please point it out for me.
> 
> Basically, here in this function
> 
> https://github.com/meshery/meshery/blob/1949742026f4828ddbd3c5bf90d5008b51550c78/server/helpers/utils/utils.go#L112
> 
> 
> we are expecting value and, based on this, deciding whether the channel is closed or not. https://github.com/meshery/meshery/pull/7309/files#diff-c9ba355fc474472a3b057eed347eb447b73f8ae534b2277c78f1767ac04a8ad0L33 and here we are publishing to the same channel.

I don't think that's an issue since the select is non-blocking, an empty struct will be received making `IsClosed` function return `true` hence the line you mention never executes. I think that's why the `IsClosed` is used.


> Secondly, we were closing channel on the receiver's side.
Good point.

**Comment by MUzairS15** [comment]
Suppose there are 2 instances of this subscription, in first instance we are signalling there is change in data and (say context switch happens ) in second instance we are checking for same channel is closed or not, since data was send to this channel by first instance the first case in select will be chosen returning true, indicating that channel is closed but in actual that channel is not closed


I might not be able to explain clearly but I can see the condition 😅, anyways I think this is better than the current implementation.

**Comment by Azanul** [comment]
> Suppose there are 2 instances of this subscription, in first instance we are signalling there is change in data and (say context switch happens ) in second instance we are checking for same channel is closed or not, since data was send to this channel by first instance the first case in select will be chosen returning true, indicating that channel is closed but in actual that channel is not closed

Wouldn't both instances have there own channels? The access to channels cross-request is only available through `r.Config.ConfigurationChannel` methods, where every channel is being checked before being pushed to.
 
> I might not be able to explain clearly but I can see the condition sweat_smile, anyways I think this is better than the current implementation.
I agree to a certain degree.

Try this once if possible (detect-data-races)[https://yourbasic.org/golang/detect-data-races/]

**Pull request closed without merging**